### PR TITLE
[Profiler] Add garbage collections in profile

### PR DESCRIPTION
**Pull request by chrisnas**
## Summary of changes
Add Stop the World periods due to garbage collections into profile samples

## Reason for change
Provide information for timeline

## Implementation details
Listen to GC_SUSPEND_EE_BEGIN/ EVENT_GC_RESTART_EE_END + other GC events to compute Stop The World duration

## Test coverage
Added tests

### CODE REVIEW

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp [hunk 1]
```diff
@@ -165,4 +271,224 @@ bool ClrEventsParser::TryGetEventInfo(LPCBYTE pMetadata, ULONG cbMetadata, WCHAR
...
+    }
+    else
+    {
+        // bad luck: a xxxBegin event has been missed
+        LOG_GC_EVENT("### missing Begin event");
+    }
+    gc.PauseDuration += suspensionDuration;
+
+    // could be the end of a gen0/gen1 or of a non concurrent gen2 GC
+    if (
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp:408]: why not having the expression on the same line ?

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp [hunk 2]
```diff
@@ -165,4 +271,224 @@ bool ClrEventsParser::TryGetEventInfo(LPCBYTE pMetadata, ULONG cbMetadata, WCHAR
...
+    // Note: last event for non background GC (will be GcGlobalHeapHistory for background gen 2)
+    GCDetails& gc = GetCurrentGC();
+    if (gc.Number == -1)
+    {
+        return;
+    }
+
+    // this is the last event for a gen0/gen1 foreground collection during a background gen2 collections
+    if (
+        (_currentBGC.Number != -1) &&
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp:439]: same

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.h [hunk 3]
```diff
@@ -47,18 +50,78 @@ struct ContentionStopV1Payload
     uint16_t ClrInstanceId;    // Unique ID for the instance of CLR.
     double_t DurationNs;       // Duration of the contention (without spinning)
 };
+
+struct GCStartPayload
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.h:54]: shouldn't those structs be in the `pragma pack` section ?

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/CollectorBase.h [hunk 4]
```diff
@@ -161,6 +161,16 @@ class CollectorBase
         ProcessID pid;
         std::string appDomainName;
 
+        // check for null AppDomainId (garbage collection for example)
+        if (rawSample.AppDomainId == 0)
+        {
+            sample.SetAppDomainName("CLR");
+            // TODO: is pid mandatory?
+            sample.SetPid("0");
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/CollectorBase.h:169]: why not setting the current process id ?

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp [hunk 5]
```diff
@@ -1,4 +1,5 @@
 // Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp:2]: duplicate line

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/CorProfilerCallback.cpp [hunk 6]
```diff
@@ -784,10 +829,14 @@ HRESULT STDMETHODCALLTYPE CorProfilerCallback::Initialize(IUnknown* corProfilerI
...
         //  - ContentionStop_V1
+        //  - GC related events
 
         UINT64 activatedKeywords = 0;
 
-        if (_pConfiguration->IsAllocationProfilingEnabled())
+        if (
+            (_pConfiguration->IsAllocationProfilingEnabled()) ||
+            (_pConfiguration->IsGarbageCollectionProfilingEnabled())
+            )
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/CorProfilerCallback.cpp:839]: put the comment here as I cannot add the comment on line 853:
Should we play with the verbosity level ?

For example allocation profiler needs the verbosity 5, but contention or GC profiler might not need this level.
We would choose the highest one depending on the enabled profilers.

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/RawStopTheWorldSample.h [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+#include "GCBaseRawSample.h"
+#include "RawSample.h"
+#include "Sample.h"
+
+class RawStopTheWorldSample : public GCBaseRawSample
+{
+public:
+    // Duration is the suspension time so default sample value
+    virtual void DoAdditionalTransform(Sample& sample, uint32_t valueOffset) const
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/RawStopTheWorldSample.h:14]: nit `override` keyword

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/Sample.cpp [hunk 8]
```diff
@@ -14,6 +14,16 @@ const std::string Sample::ExceptionTypeLabel = "exception type";
...
 const std::string Sample::AllocationClassLabel = "allocation class";
 
+// garbage collection related labels
+const std::string Sample::TimelineEventTypeLabel = "event";
+    const std::string Sample::TimelineEventTypeStopTheWorld = "stw";
+    const std::string Sample::TimelineEventTypeGarbageCollection = "gc";
+        const std::string Sample::GarbageCollectionReasonLabel = "gc reason";   // look at GCReason enumeration
+        const std::string Sample::GarbageCollectionTypeLabel = "gc type";       // look at GCType enumeration
+        const std::string Sample::GarbageCollectionCompactingLabel = "gc compacting"; // true or false
+const std::string Sample::GarbageCollectionGenerationLabel = "gc generation";
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/Sample.cpp:24]: indentation ?

File: profiler/test/Datadog.Profiler.IntegrationTests/GarbageCollections/GarbageCollectionsProfilerTest.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+        private readonly ITestOutputHelper _output;
+
+        public GarbageCollectionsProfilerTest(ITestOutputHelper output)
+        {
+            _output = output;
+        }
+
+        [TestAppFact("Samples.Computer01", new[] { "net6.0" })]
+        public void ShouldGetAllocationSamples(string appName, string framework, string appAssembly)
```
Comment (gleocadie) [inline profiler/test/Datadog.Profiler.IntegrationTests/GarbageCollections/GarbageCollectionsProfilerTest.cs:32]: the name does not reflect the test :^)

**Review by gleocadie — APPROVED** [review]
comments there and there, otherwise looks good

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/CollectorBase.h (hunk 4 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/CollectorBase.h:169]: The current implementation relies on a call to ICorProfilerInfo that requires the AppDomainID that we don't have

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp (hunk 5 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp:2]: Good catch  :^)

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/Sample.cpp (hunk 8 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/Sample.cpp:24]: I tried to show the relations between the different labels...

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp (hunk 1 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp:408]: ✅

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/RawStopTheWorldSample.h (hunk 7 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/RawStopTheWorldSample.h:14]: ✅  Good catch!

File: profiler/test/Datadog.Profiler.IntegrationTests/GarbageCollections/GarbageCollectionsProfilerTest.cs (hunk 9 above)
Comment (chrisnas) [reply profiler/test/Datadog.Profiler.IntegrationTests/GarbageCollections/GarbageCollectionsProfilerTest.cs:32]: ✅ done

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/CorProfilerCallback.cpp (hunk 6 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/CorProfilerCallback.cpp:839]: ✅ Good idea (done)

**Pull request merged**